### ISSUE TITLE
SqDistAlgorithm.process_one() problem with empty trace

### ISSUE THREAD

**Issue by erigler-usgs** [usgs/geomag-algorithms#239]
When SqDistAlgorithm.additive() (the heart of SqDistAlgorithm) receives a zero-length array-like input, it returns a zero-length array-like object, and correctly leaves all initial values unchanged. However, the wrapper SqDistAlgorithm.process_one() incorrectly updates the `next_starttime` state parameter.

https://github.com/usgs/geomag-algorithms/blob/811adb911741fea7b8bb70a9b992804b8682dff6/geomagio/algorithm/SqDistAlgorithm.py#L261

This is due to weirdness in the obspy.core.Trace class, wherein you can set the number of points to zero, but then the `Trace.stats.starttime` and `Trace.stats.endtime` are not well-defined.

https://github.com/usgs/geomag-algorithms/blob/811adb911741fea7b8bb70a9b992804b8682dff6/geomagio/TimeseriesUtility.py#L384-L388

The solution I propose is to check the length of the array returned from SqDistAlgorithm.additive(), and multiply this by `delta` before adding to the existing `next_starttime` state parameter. This is a bit of a kludge, but I would argue that the obspy.Trace data model is just not sufficient to deal with this situation, and we should not waste effort to try to get it to do the right thing.

@jmfee-usgs, I'd appreciate your take on this before I make this change and push out a new release. FYI, the practical problem this causes is that `next_starttime` gets incremented even when no processing is done because the input time series got trimmed to zero-length to accommodate the `realtime_gap` calculated in the Controller.

**Comment by jmfee-usgs** [comment]
Would it be easier to skip process_one when `trace.npts == 0`?

Alternatively, the calculation could use npts instead of adding to endtime:
```
self.next_starttime = trace.stats.starttime + (trace.stats.delta * trace.stats.npts)
```

**Comment by erigler-usgs** [comment]
A reason **not** to skip process_one() is that the additive function has the ability to forecast into the future. Actually, there are two ways it can do this: One time-shifts the inputs vs. outputs using the `hstep` parameter; and one actually simulates `fc` discrete steps beyond the last input by propagating the state forward in time. 

We don't use either of these currently, and probably need to document this functionality better before making it available through the Controller, but I'd rather not make such future changes more difficult than they need to be.

So, this requires that we do **not** rely on trace.stats.npts. What I think we need is:

```python
self.next_starttime = self.next_starttime + trace.stats.delta * (len(yhat) + self.hstep)
```

I'll proceed with this unless you notice a flaw in my logic. Thanks for your feedback.

**Comment by jmfee-usgs** [comment]
I agree `additive` may forecast into the future, but not sure if `process`/`process_one` would under normal usage...  For the "realtime" equivalent, if the Algorithm needs to project into the future: the Controller will pass a non-empty (null data values) trace for the requested samples.  Does that change your answer, or do you still think yhat/hstep is needed?

**Comment by erigler-usgs** [comment]
@jmfee-usgs: bear with me while I risk repeating what you may have already said to me offline...I just want to get this down so there is something to refer to in the future when I'm trying to find the mental space I'm in right now (and where I think you and I are mostly in agreement):

Right now, if one were to use the `fc` parameter to simulate fc steps beyond the last passed input, this would also propagate the state fc steps. So, for the zero-length input (`yobs`) example, if `fc=5`, the output (`yhat`) will be 5 in length, and the state (`yhat0, l0, b0 s0`, etc.) will be consistent with 5 steps beyond the last input value in `yobs`.

Now, your proposal would not update `next_starttime` at all, and on reflection, I agree this is what we want since the point of the "realtime" logic is to be able to pick up processing as new inputs become available.

But, if we want to allow the fc parameter to be used, we need to modify additive() to return the state as of the last processed input, NOT the state associated with the fc-step simulation. Since `hstep` is just a time-shift, and should never lead to more outputs than input anyway, there is no problem here.

One remaining problem is that we now need to address what happens to the updated Traces constructed from additive()'s output and returned by process_one() when `fc` or `hstep` are used. That shouldn't be too difficult.

So, in summary:

1. we cannot bypass the call to additive if npts is zero because of `fc`;
2. we should apply your above proposal for adjusting `next_starttime`; but...
3. we should modify additive() to return l0, yhat0, etc. consistent with the last input, and NOT with the last output;
4. we do NOT need to change anything to deal with hstep, since the state should be aligned with the inputs, and not the outputs.
5. we should modify the post-processing Trace construction in process_one() to deal with `fc` and `hstep` accordingly (adding and/or shifting time steps).
